discovery_engine_core/bindings/src/types/duration.rs

```diff
@@ -0,0 +1,99 @@
+// Copyright 2022 Xayn AG
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, version 3.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+//! FFI functions for handling `Duration` fields.
+
+use std::time::Duration;
+
+/// Initializes a `Duration` field at given place.
+///
+/// # Safety
+///
+/// It must be valid to write a [`Duration`] to given pointer.
+#[no_mangle]
+pub unsafe extern "C" fn init_duration_at(place: *mut Duration, seconds: u64, nanos: u32) {
+    unsafe {
+        place.write(Duration::new(seconds, nanos));
+    }
+}
+
+/// Gets the seconds of a duration at given place.
+///
+/// # Safety
+///
+/// The pointer must point to a valid [`Duration`] instance.
+#[no_mangle]
+pub unsafe extern "C" fn get_duration_seconds(place: *mut Duration) -> u64 {
+    unsafe { &*place }.as_secs()
+}
+
+/// Gets the (subseconds) nanoseconds of a duration at given place.
+///
+/// # Safety
+///
+/// The pointer must point to a valid [`Duration`] instance.
+#[no_mangle]
+pub unsafe extern "C" fn get_duration_nanos(place: *mut Duration) -> u32 {
+    unsafe { &*place }.subsec_nanos()
+}
+
+/// Alloc an uninitialized `Box<Duration>`, mainly used for testing.
+#[cfg(feature = "additional-ffi-methods")]
+#[no_mangle]
+pub extern "C" fn alloc_uninitialized_duration_box() -> *mut Duration {
```

Comment:
We can drop the `box` suffix.

Reply:
but we alloc a box of an `Duration`, not something else.

Reply:
But the user cannot see it. They get a pointer to `Duration` and they don't care how it was allocated.

Reply:
They do have to care, because if this is returned to dart (by a rust ffi function) dart still needs to call the right drop method (because we returned it to dart).

And who says that we don't maybe at some point also return some `Arc<T>`'s in some future API in which case for some types there would be two drop methods both taking a `*mut T`.

Reply:
Now `alloc_`/`drop_` imply box, so the prefix is removed.